File: src/core/expr/code/impl.rb
require 'core/expr/code/eval'
require 'core/expr/code/lvalue'
require 'core/semantics/code/interpreter'
require 'core/expr/code/env'

module Impl
  #note that the closure stores variable states only,
  #not interpreter state
  #so calling a closure may produce different behavior
  #depending on where it is evaluated because the
  #interpreter may be different
  class Closure
    attr_reader :formals, :body

    def self.make_closure(body, formals, env, interp)
      Closure.new(body, formals, env, interp).method('call_closure')
    end

    def initialize(body, formals, env, interp)
      @body = body
      @formals = formals
      @env = env
      @interp = interp
    end

    #params are the values used to call this function
    #args are used by the interpreter
    def call_closure(*params)
      #puts "CALL #{@formals} #{params}"
      nv = {}
      @formals.each_with_index do |f,i|
        nv[f] = params[i]
      end
      nenv = Env::HashEnv.new(nv, @env)
      @interp.dynamic_bind(env: nenv) do
        @interp.eval(@body)
      end
    end

    def to_s()
      "#<Closure(#{@formals.map{|f|f.name}.join(", ")}) {#{@body}}>"
    end
  end

  module EvalCommand
  
    include Eval::EvalExpr
    include Lvalue::LValueExpr
    
    include Interpreter::Dispatcher    
      
    def eval(obj)
      dispatch_obj(:eval, obj)
    end
    
    def eval_EWhile(obj)
      while eval(obj.cond)
        eval(obj.body)
      end
    end

    def eval_EFor(obj)
      env = {}
      env[obj.var] = nil
      nenv = Env::HashEnv.new(env, @D[:env])
      eval(obj.list).each do |val|
        nenv[obj.var] = val
        dynamic_bind(env: nenv) do
          eval(obj.body)
        end
      end
    end
  
    def eval_EIf(obj)
      if eval(obj.cond)
        eval(obj.body)
      elsif !obj.body2.nil?
        eval(obj.body2)
      end
    end

    def eval_EBlock(obj)
      res = nil
      #fundefs are able to see each other but not any other variable created in the block
      defenv = Env::HashEnv.new({}, @D[:env])
      dynamic_bind(in_fc: false, env: defenv) do
        obj.fundefs.each do |c|
          eval(c)
        end
      end
      #rest of body can see fundefs
      env1 = Env::HashEnv.new({}, defenv)
      dynamic_bind(in_fc: false, env: env1) do
        obj.body.each do |c|
          res = eval(c)
        end
      end
      res
    end

    def eval_EFunDef(obj)
      forms = []
      obj.formals.each {|f| forms << f.name}
      @D[:env][obj.name] = Impl::Closure.make_closure(obj.body, forms, @D[:env], self)
      nil
    end

    def eval_ELambda(obj)
      forms = []
      obj.formals.each {|f| forms << f.name}
      Proc.new { |*p| Impl::Closure.make_closure(obj.body, forms, @D[:env], self).call(*p) }
    end

    def eval_EFunCall(obj)
      m = dynamic_bind(in_fc: true) do
        eval(obj.fun)
      end
      if obj.lambda.nil?
        
        m.call(*(obj.params.map{|p|eval(p)}))
      else
        b = eval(obj.lambda)
        m.call(*(obj.params.map{|p|eval(p)}), &b) 
      end
    end

    def eval_EAssign(obj)
      lvalue(obj.var).value = eval(obj.val)
    end
  end
  
  class EvalCommandC
    include EvalCommand
    def initialize
    end
  end
        
  def self.eval(obj, args = {env:{}})

  	local_funs = {}
  	local_funs["MAX"] = Proc.new { |*a|
	  	  max = nil
	  	  #puts "MAX #{a}"
	  	  a.each do |b|
		  	  if b.is_a?(Enumerable)
		  	    b.each do |v|
		  	      max = max.nil? ? v : [max, v].max
		  	    end
		  	  else
		  	    max = max.nil? ? b : [max, b].max
		  	  end
		  	end
		  	max
	    }
  	local_funs["MIN"] = Proc.new { |*a|
	  	  min = nil
	  	  #puts "MIN #{a}"
	  	  a.each do |b|
		  	  if b.is_a?(Enumerable)
		  	    b.each do |v|
		  	      min = min.nil? ? v : [min, v].min
		  	    end
		  	  else
		  	    min = min.nil? ? b : [min, b].min
		  	  end
		  	end
		  	min
      }
  	local_funs["SUM"] = Proc.new { |*a|
        sum = 0
	  	  a.each do |b|
	  	    if b.is_a?(Enumerable)
		  	    b.each do |v|
		  	      sum = sum + v
		  	    end
		  	  else
		  	    sum = sum + b
		  	  end
		  	end
		  	sum
	    }
  	local_funs["COUNT"] = Proc.new { |*a|
	  	  count = nil
	  	  a.each do |b|
		  	  if b.is_a?(Enumerable)
		  	    b.each do |v|
		  	      count = count.nil? ? 1 : count + 1
		  	    end
		  	  else
		  	    count = count.nil? ? 1 : count + 1
		  	  end
		  	end
		  	count
	    }
  	local_funs["AVERAGE"] = Proc.new { |*a|
	  	  local_funs["SUM"].call(*a) / local_funs["COUNT"].call(*a)
	    }
  	local_funs["STDEV"] = Proc.new { |*a|
	  	  average = local_funs["AVERAGE"].call(*a)
	  	  sumvariance = 0
	  	  count = 0
	  	  a.each do |b|
		  	  if b.is_a?(Enumerable)
		  	    b.each do |v|
		  	      p = v - average
		  	      sumvariance = sumvariance + (p * p) 
		  	      count = count + 1
		  	    end
		  	  else
	  	      p = b - average
	  	      sumvariance = sumvariance + (p * p)
    	      count = count + 1
		  	  end
		  	end
		  	Math.sqrt(sumvariance / count)
	    }
  	local_funs["MEDIAN"] = Proc.new { |*a|
	  	  arr = []
	  	  a.each do |b|
		  	  if b.is_a?(Enumerable)
		  	    b.each do |v|
		  	      arr << v
		  	    end
		  	  else
		  	    arr << b
		  	  end
		  	end
			  mid = arr.length / 2
			  sorted = arr.sort
			  #puts "MID #{mid} #{sorted}"
			  mid.odd? ? sorted[mid] : 0.5 * (sorted[mid] + sorted[mid - 1])
	    }
    
  
    nv = Env::HashEnv.new(local_funs, args[:env])
    args[:env] = nv
    interp = EvalCommandC.new
    interp.dynamic_bind(args) do
      interp.eval(obj)
    end
  end
end

File: src/js/core/expr/code/impl.js
'use strict'

//// Impl ////

var cwd = process.cwd() + '/';
var Eval = require(cwd + "core/expr/code/eval.js");
var Lvalue = require(cwd + "core/expr/code/lvalue.js");
var Interpreter = require(cwd + "core/semantics/code/interpreter.js");
var Env = require(cwd + "core/expr/code/env.js");
var Enso = require(cwd + "enso.js");

var Impl;

var eval_M = function(obj, args = Enso.EMap.new({env: Enso.EMap.new()})) {
  var self = this, local_funs, max, min, sum, count, average, sumvariance, p, arr, mid, sorted, nv, interp;
  local_funs = Enso.EMap.new();
  local_funs .set$("MAX", function(...a) {
    max = null;
    a.each(function(b) {
      if (Enso.System.test_type(b, Enumerable)) {
        return b.each(function(v) {
          return max = max == null
            ? v
            : [max, v].max();
        });
      } else {
        return max = max == null
          ? b
          : [max, b].max();
      }
    });
    return max;
  });
  local_funs .set$("MIN", function(...a) {
    min = null;
    a.each(function(b) {
      if (Enso.System.test_type(b, Enumerable)) {
        return b.each(function(v) {
          return min = min == null
            ? v
            : [min, v].min();
        });
      } else {
        return min = min == null
          ? b
          : [min, b].min();
      }
    });
    return min;
  });
  local_funs .set$("SUM", function(...a) {
    sum = 0;
    a.each(function(b) {
      if (Enso.System.test_type(b, Enumerable)) {
        return b.each(function(v) {
          return sum = sum + v;
        });
      } else {
        return sum = sum + b;
      }
    });
    return sum;
  });
  local_funs .set$("COUNT", function(...a) {
    count = null;
    a.each(function(b) {
      if (Enso.System.test_type(b, Enumerable)) {
        return b.each(function(v) {
          return count = count == null
            ? 1
            : count + 1;
        });
      } else {
        return count = count == null
          ? 1
          : count + 1;
      }
    });
    return count;
  });
  local_funs .set$("AVERAGE", function(...a) {
    return local_funs.get$("SUM")(...a) / local_funs.get$("COUNT")(...a);
  });
  local_funs .set$("STDEV", function(...a) {
    average = local_funs.get$("AVERAGE")(...a);
    sumvariance = 0;
    count = 0;
    a.each(function(b) {
      if (Enso.System.test_type(b, Enumerable)) {
        return b.each(function(v) {
          p = v - average;
          sumvariance = sumvariance + p * p;
          return count = count + 1;
        });
      } else {
        p = b - average;
        sumvariance = sumvariance + p * p;
        return count = count + 1;
      }
    });
    return Math.sqrt(sumvariance / count);
  });
  local_funs .set$("MEDIAN", function(...a) {
    arr = [];
    a.each(function(b) {
      if (Enso.System.test_type(b, Enumerable)) {
        return b.each(function(v) {
          return arr.push(v);
        });
      } else {
        return arr.push(b);
      }
    });
    mid = arr.size_M() / 2;
    sorted = arr.sort();
    if (mid.odd_P()) {
      return sorted.get$(mid);
    } else {
      return 0.5 * (sorted.get$(mid) + sorted.get$(mid - 1));
    }
  });
  nv = Env.HashEnv.new(local_funs, args.get$("env"));
  args .set$("env", nv);
  interp = EvalCommandC.new();
  return interp.dynamic_bind(function() {
    return interp.eval_M(obj);
  }, args);
};

class Closure extends Enso.EnsoBaseClass {
  static new(...args) { return new Closure(...args) };

  static make_closure(body, formals, env, interp) {
    var self = this;
    return Closure.new(body, formals, env, interp).method("call_closure");
  };

  formals() { return this.formals$ };

  body() { return this.body$ };

  constructor(body, formals, env, interp) {
    super();
    var self = this;
    self.body$ = body;
    self.formals$ = formals;
    self.env$ = env;
    self.interp$ = interp;
  };

  call_closure(...params) {
    var self = this, nv, nenv;
    nv = Enso.EMap.new();
    self.formals$.each_with_index(function(f, i) {
      return nv .set$(f, params.get$(i));
    });
    nenv = Env.HashEnv.new(nv, self.env$);
    return self.interp$.dynamic_bind(function() {
      return self.interp$.eval_M(self.body$);
    }, Enso.EMap.new({env: nenv}));
  };

  to_s() {
    var self = this;
    return Enso.S("#<Closure(", self.formals$.map(function(f) {
      return f.name();
    }).join(", "), ") {", self.body$, "}>");
  };
};

function EvalCommand(parent) {
  return class extends Enso.mix(parent, Eval.EvalExpr, Lvalue.LValueExpr, Interpreter.Dispatcher) {
    eval_M(obj) {
      var self = this;
      return self.dispatch_obj("eval", obj);
    };

    eval_ESequence(obj) {
      var self = this;
      return obj.items().each(function(block) {
        return self.eval_M(block);
      });
    };

    eval_EWhile(obj) {
      var self = this;
      while (self.eval_M(obj.cond())) {
        self.eval_M(obj.body());
      }
    };

    eval_EFor(obj) {
      var self = this, env, nenv;
      env = Enso.EMap.new();
      env .set$(obj.var(), null);
      nenv = Env.HashEnv.new(env, self.D$.get$("env"));
      return self.eval_M(obj.list()).each(function(val) {
        nenv .set$(obj.var(), val);
        return self.dynamic_bind(function() {
          return self.eval_M(obj.body());
        }, Enso.EMap.new({env: nenv}));
      });
    };

    eval_EIf(obj) {
      var self = this;
      if (self.eval_M(obj.cond())) {
        return self.eval_M(obj.body());
      } else if (! (obj.body2() == null)) {
        return self.eval_M(obj.body2());
      }
    };

    eval_EBlock(obj) {
      var self = this, res, defenv, env1;
      res = null;
      defenv = Env.HashEnv.new(Enso.EMap.new(), self.D$.get$("env"));
      self.dynamic_bind(function() {
        return obj.fundefs().each(function(c) {
          return self.eval_M(c);
        });
      }, Enso.EMap.new({in_fc: false, env: defenv}));
      env1 = Env.HashEnv.new(Enso.EMap.new(), defenv);
      self.dynamic_bind(function() {
        return obj.body().each(function(c) {
          return res = self.eval_M(c);
        });
      }, Enso.EMap.new({in_fc: false, env: env1}));
      return res;
    };

    eval_EFunDef(obj) {
      var self = this, forms;
      forms = [];
      obj.formals().each(function(f) {
        return forms.push(f.name());
      });
      self.D$.get$("env") .set$(obj.name(), Impl.Closure.make_closure(obj.body(), forms, self.D$.get$("env"), self));
      return null;
    };

    eval_ELambda(obj) {
      var self = this, forms;
      forms = [];
      obj.formals().each(function(f) {
        return forms.push(f.name());
      });
      return function(...p) {
        return Impl.Closure.make_closure(obj.body(), forms, self.D$.get$("env"), self)(...p);
      };
    };

    eval_EFunCall(obj) {
      var self = this, m, b;
      m = self.dynamic_bind(function() {
        return self.eval_M(obj.fun());
      }, Enso.EMap.new({in_fc: true}));
      if (obj.lambda() == null) {
        return m(...obj.params().map(function(p) {
          return self.eval_M(p);
        }));
      } else {
        b = self.eval_M(obj.lambda());
        return m(b, ...obj.params().map(function(p) {
          return self.eval_M(p);
        }));
      }
    };

    eval_EAssign(obj) {
      var self = this;
      return self.lvalue(obj.var()).set_value(self.eval_M(obj.val()));
    }; }};

class EvalCommandC extends Enso.mix(Enso.EnsoBaseClass, EvalCommand) {
  static new(...args) { return new EvalCommandC(...args) };

  constructor() {
    super();
  };
};

Impl = {
  eval_M: eval_M,
  Closure: Closure,
  EvalCommand: EvalCommand,
  EvalCommandC: EvalCommandC,
};
module.exports = Impl ;
